### PR TITLE
Add glob support

### PR DESCRIPTION
**Pull request by thiagodelgado111**
- [x] Add glob support using minimatch
- [x] Add unit tests
- [x] Add precommit-hook

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 .npm
 .coverage
+node_modules
```
Comment (serut) [inline .gitignore]: this is required ? in the package there are a .npm folder, no node_modules from what I see

File: client/methods.js [hunk 2]
```diff
@@ -1,9 +1,11 @@
 import { $ } from 'meteor/jquery';
+import { Meteor } from 'meteor/meteor';
```
Comment (serut) [inline client/methods.js:2]: Cool improvement. Need to check if istanbul can still access to these methods with Package...

File: client/methods.js [hunk 3]
```diff
@@ -42,21 +46,22 @@ Package['meteor']['Meteor'].sendCoverage = function (callback) {
...
     $.ajax({
         method: 'GET',
         url: url,
         success: function(data) {
             try {
                 let result = JSON.parse(data);
-                if (result.type === "success") {
-                    return callback();
+                if (result.type !== "success") {
+                  throw new Error('An unexpected error occurred while trying to export coverage data');
```
Comment (serut) [inline client/methods.js:88]: callback is not called? so the client caller (spacejam, ...) cannot end correctly. Too bad

File: client/methods.tests.js [hunk 4]
```diff
@@ -1,136 +1,66 @@
...
-                function(err) {
-                    assert.isUndefined(err);
-                    done();
-                }
-            );
-        } catch (e) {
-            console.error (e, e.stack)
-            done(e);
-        }
+        expect(callback).to.have.been.deep.calledWith({SUCCESS: 0, FAILED: 0, TOTAL: 0});
```
Comment (serut) [inline client/methods.tests.js:29]: Strange value no ?

File: conf/default-coverage.json [hunk 5]
```diff
@@ -1,79 +1,26 @@
...
-        },
-        "serverside": [
-            "livedata.js",
-            "hot-code-push.js",
-            "launch-screen.js",
-            "templating.js",
-            "autoupdate.js",
-            "reload.js",
-            "global-imports.js"
+    "include": [
```
Comment (serut) [inline conf/default-coverage.json:2]: What is this ? How do you force to import a file ?

File: conf/default-coverage.json [hunk 6]
```diff
@@ -1,79 +1,26 @@
...
+        "**/lib/**/methods.js"
+    ],
+    "exclude": {
+        "general": [
+            "/**/packages/**/*",
+            "/**/*test.js",
+            "../**/packages/**/*",
+            "/**/lib/test/**/*",
+            "/**/*stream*.js",
```
Comment (serut) [inline conf/default-coverage.json:11]: too much general, cannot be a default value

File: server/context/conf.js [hunk 7]
```diff
@@ -1,54 +1,45 @@
...
-            configuration.ignore.others = defautConf.ignore.others;
-        }
-        if (configuration.output === undefined) {
-            Log.info("Loading default configuration: others");
-            configuration.output = defautConf.output;
-        }
-    } else {
-        Log.info("Loading default configuration");
-        configuration = defautConf;
+
```
Comment (serut) [inline server/context/conf.js:24]: So here you didn't look if exclude.server is defined or not. So there are no default value if the user define half of the schema, like it was previously

File: server/handlers.js [hunk 8]
```diff
@@ -1,8 +1,10 @@
-import Instrumenter from './services/instrumenter'
```
Comment (serut) [inline server/handlers.js:1]: Remove this file from pull request.

File: server/main.js [hunk 9]
```diff
@@ -2,8 +2,8 @@ import Instrumenter from './services/instrumenter';
 import CoverageData from './services/coverage-data';
```
Comment (serut) [inline server/main.js:2]: Remove this file from pull request.

File: server/report/report-commun.js [hunk 10]
```diff
@@ -1,6 +1,6 @@
 import fs from 'fs';
```
Comment (serut) [inline server/report/report-commun.js:1]: Remove this file from pull request.

File: server/report/report-coverage.js [hunk 11]
```diff
@@ -1,12 +1,11 @@
-import Conf from '../context/conf'
```
Comment (serut) [inline server/report/report-coverage.js:1]: Remove this file from pull request.

File: server/report/report-generic.js [hunk 12]
```diff
@@ -1,8 +1,8 @@
-import CoverageData from './../services/coverage-data'
```
Comment (serut) [inline server/report/report-generic.js:1]: Remove this file from pull request.

File: server/report/report-html.js [hunk 13]
```diff
@@ -1,9 +1,9 @@
-import Conf from '../context/conf'
```
Comment (serut) [inline server/report/report-html.js:1]: Remove this file from pull request.

File: server/report/report-http.js [hunk 14]
```diff
@@ -1,6 +1,6 @@
-import CoverageData from '../services/coverage-data'
```
Comment (serut) [inline server/report/report-http.js:1]: Remove this file from pull request.

File: server/report/report-json-summary.js [hunk 15]
```diff
@@ -1,9 +1,9 @@
-import Conf from './../context/conf'
```
Comment (serut) [inline server/report/report-json-summary.js:1]: Remove this file from pull request.

File: server/report/report-service.js [hunk 16]
```diff
@@ -1,12 +1,12 @@
-import IstanbulFile from './report-generic'
```
Comment (serut) [inline server/report/report-service.js:1]: Remove this file from pull request.

File: server/report/report-teamcity.js [hunk 17]
```diff
@@ -1,9 +1,9 @@
-import Conf from './../context/conf'
```
Comment (serut) [inline server/report/report-teamcity.js:1]: Remove this file from pull request.

File: server/report/report-text-summary.js [hunk 18]
```diff
@@ -1,9 +1,9 @@
-import Conf from '../context/conf'
```
Comment (serut) [inline server/report/report-text-summary.js:1]: Remove this file from pull request.

File: server/router.js [hunk 19]
```diff
@@ -1,5 +1,5 @@
-import Handlers from './handlers'
```
Comment (serut) [inline server/router.js:1]: Remove this file from pull request.

File: server/services/core.js [hunk 20]
```diff
@@ -1,4 +1,4 @@
-import Conf from './../context/conf'
```
Comment (serut) [inline server/services/core.js:1]: Remove this file from pull request.

File: server/services/instrumenter.js [hunk 21]
```diff
@@ -1,11 +1,13 @@
...
-import Conf from './../context/conf'
-let hookLoader, instrumentJs, checkIfAutorised, shallInstrumentClientScript, shallInstrumentServerScript
+import Conf from './../context/conf';
+
+let hookLoader, instrumentJs, shouldIgnore, shallInstrumentClientScript, shallInstrumentServerScript;
 
 if (Conf.IS_COVERAGE_ACTIVE) {
     //var im = Npm.require('istanbul-middleware');
     var istanbulAPI = Npm.require('istanbul-api'),
+        minimatch = Npm.require('minimatch'),
```
Comment (serut) [inline server/services/instrumenter.js]: These npm.require can be transformed into import minimatch from "minimatch" right ?

File: server/services/instrumenter.js [hunk 22]
```diff
@@ -39,30 +41,49 @@ if (Conf.IS_COVERAGE_ACTIVE) {
...
+         shouldIgnore = Conf.exclude.server.some(pattern => fileMatch(pattern));
+       }
+
+       if (!shouldIgnore && Conf.exclude.client && !isAServerSideFile) {
+         shouldIgnore = Conf.exclude.client.some(pattern => fileMatch(pattern));
+       }
+
+       return shouldIgnore;
+
+       function fileMatch(pattern) {
```
Comment (serut) [inline server/services/instrumenter.js]: Is is mandatory to declare this function at that level ?

File: server/services/source-map.js [hunk 23]
```diff
@@ -1,8 +1,8 @@
 import Log from './../context/log';
```
Comment (serut) [inline server/services/source-map.js]: Remove file from pull request

File: server/tests.js [hunk 24]
```diff
@@ -1,26 +1,35 @@
 import MeteorCoverage from 'meteor/lmieulet:meteor-coverage';
```
Comment (serut) [inline server/tests.js]: replace into import {Instrumenter} from 'meteor/lmieulet:meteor-coverage';

File: .gitignore (hunk 1 above)
Comment (thiagodelgado111) [inline .gitignore]: Yes it is, I added a package.json with a few scripts to run, I've also installed eslint and precommit-hook as devDependencies

File: client/methods.js (hunk 2 above)
Comment (thiagodelgado111) [inline client/methods.js:2]: e2e tests would guarantee that but for unit tests, it's already working fine

File: client/methods.js (hunk 3 above)
Comment (thiagodelgado111) [inline client/methods.js:88]: The callback should be called if `result.type` is === to `success`, right?

File: client/methods.tests.js (hunk 4 above)
Comment (thiagodelgado111) [inline client/methods.tests.js:29]: No, this is testing for an empty `global['__coverage__']`(https://github.com/thiagodelgado111/meteor-coverage/blob/0bc69a18547d19660a7e7bdf281df4f7c2d0e7bd/client/methods.js#L24)

File: conf/default-coverage.json (hunk 5 above)
Comment (thiagodelgado111) [inline conf/default-coverage.json:2]: I'm not enforcing it to be imported, I'm adding it as an exception for say, ignored folders. So if a filepath matches this pattern it is never ignored even if it match a `ignore` pattern rule

File: conf/default-coverage.json (hunk 6 above)
Comment (thiagodelgado111) [inline conf/default-coverage.json:11]: Agreed, do you have any suggestions to leave as default?

File: server/context/conf.js (hunk 7 above)
Comment (thiagodelgado111) [inline server/context/conf.js:24]: I see but I think that users are aware of this, also, what if the user don't wanna to ignore anything from client-side/server-side? 

BTW, I think that it'd be really hard to come up with a default configuration that would match every project needs, so users would end up needing to define it anyway

File: server/services/instrumenter.js (hunk 21 above)
Comment (thiagodelgado111) [inline server/services/instrumenter.js]: Yes! Good catch!

File: server/services/instrumenter.js (hunk 22 above)
Comment (thiagodelgado111) [inline server/services/instrumenter.js]: It's not a problem as I see, this is not a function that should be accessible to the other functions, I think it should be very fine

File: server/tests.js (hunk 24 above)
Comment (thiagodelgado111) [inline server/tests.js]: Yep, good catch!

File: package.json [hunk 25]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "meteor-coverage",
+  "version": "0.9.1",
+  "description": "add istanbul code instrumentation to meteor code so we can have coverage analysis",
+  "main": "package.js",
+  "scripts": {
+    "lint": "eslint .",
+    "test": "COVERAGE=1 COVERAGE_APP_FOLDER=./ COVERAGE_VERBOSE=1 meteor test-packages --once --driver-package dispatch:mocha-phantomjs ./"
```
Comment (serut) [inline package.json]: full path required for COVERAGE_APP_FOLDER ! I don't think such path will work all the time in every case.

**Pull request merged**